pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -117,88 +122,199 @@ func (n *NetworkPolicyController) deleteCNP(old interface{}) {
 // instance to the caller wherein, it will be either stored as a new Object
 // in case of ADD event or modified and store the updated instance, in case
 // of an UPDATE event.
 func (n *NetworkPolicyController) processClusterNetworkPolicy(cnp *crdv1alpha1.ClusterNetworkPolicy) *antreatypes.NetworkPolicy {
-	appliedToPerRule := len(cnp.Spec.AppliedTo) == 0
-	// appliedToGroupNames tracks all distinct appliedToGroups referred to by the ClusterNetworkPolicy,
+
+	hasPerNamespaceRule := hasPerNamespaceRule(cnp)
+	// If one of the ACNP rule is a per-namespace rule (a peer in that rule has namspaces.Match set
+	// to Self), the policy will need to be converted appliedTo per rule policy, as the appliedTo will
+	// be different for rules created for each namespace.
+	appliedToPerRule := len(cnp.Spec.AppliedTo) == 0 || hasPerNamespaceRule
+
+	// atgNamesSet tracks all distinct appliedToGroups referred to by the ClusterNetworkPolicy,
 	// either in the spec section or in ingress/egress rules.
 	// The span calculation and stale appliedToGroup cleanup logic would work seamlessly for both cases.
-	appliedToGroupNamesSet := sets.String{}
-	// Create AppliedToGroup for each AppliedTo present in ClusterNetworkPolicy spec.
-	for _, at := range cnp.Spec.AppliedTo {
-		var atg string
-		if at.Group != "" {
-			atg = n.processAppliedToGroupForCG(at.Group)
-		} else {
-			atg = n.createAppliedToGroup("", at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector)
-		}
-		if atg != "" {
-			appliedToGroupNamesSet.Insert(atg)
-		}
-	}
-	rules := make([]controlplane.NetworkPolicyRule, 0, len(cnp.Spec.Ingress)+len(cnp.Spec.Egress))
-	// Compute NetworkPolicyRule for Ingress Rule.
-	for idx, ingressRule := range cnp.Spec.Ingress {
-		// Set default action to ALLOW to allow traffic.
-		services, namedPortExists := toAntreaServicesForCRD(ingressRule.Ports)
-		var appliedToGroupNamesForRule []string
-		// Create AppliedToGroup for each AppliedTo present in the ingress rule.
-		for _, at := range ingressRule.AppliedTo {
-			var atGroup string
-			if at.Group != "" {
-				atGroup = n.processAppliedToGroupForCG(at.Group)
-			} else {
-				atGroup = n.createAppliedToGroup("", at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector)
-			}
-			if atGroup != "" {
-				appliedToGroupNamesForRule = append(appliedToGroupNamesForRule, atGroup)
-				appliedToGroupNamesSet.Insert(atGroup)
+	atgNamesSet := sets.String{}
+
+	// affectedNamespaceSelectors tracks all the appliedTo's namespaceSelectors of per-namespace rules.
+	// It is used as an index so that Namespace updates can trigger corresponding rules
+	// to re-calculate affected Namespaces.
+	var affectedNamespaceSelectors []labels.Selector
+
+	// If appliedTo is set at spec level and the ACNP has per-namespace rules, then each appliedTo needs
+	// to be split into appliedToGroups for each of its affected Namespace.
+	var clusterAppliedToAffectedNS []string
+	var atgForNamespace []string
```

Comment:
Could you add a comment for this variable? Does it mean the appliedToGroups splitted by namespaces?

Reply:
Done